elastic/kibana
# Adds error from es call to nodes.info to the nodes version compatibility response message

## Summary

Resolves https://github.com/elastic/kibana/issues/97787

- Appends the request error message to `NodesVersionCompatibility` message if we get an error from Elasticsearch
- Adds the request error itself to the Observable for services and other solutions to use if they want to
- Adds a condition to `compareNodes` that checks if the error message has changed before triggering a new emission.

### Plugin API Changes

The `NodesVersionCompatibility` interface has changed and now includes an optional  `nodesInfroRequestError` parameter:

Before: 
```ts
export interface NodesVersionCompatibility {
  isCompatible: boolean;
  message?: string;
  incompatibleNodes: NodeInfo[];
  warningNodes: NodeInfo[];
  kibanaVersion: string;
}
```

After:
```ts
export interface NodesVersionCompatibility {
  isCompatible: boolean;
  message?: string;
  incompatibleNodes: NodeInfo[];
  warningNodes: NodeInfo[];
  kibanaVersion: string;
  nodesInfoRequestError?: Error // any error that's thrown from a call to `nodes.info` will surface here.
}
```

If present, the error is also available on `ElasticsearchStatusMeta`:
Before:
```ts
export interface ElasticsearchStatusMeta {
  warningNodes: NodesVersionCompatibility['warningNodes'];
  incompatibleNodes: NodesVersionCompatibility['incompatibleNodes'];
}
```

After:
```ts
export interface ElasticsearchStatusMeta {
  warningNodes: NodesVersionCompatibility['warningNodes'];
  incompatibleNodes: NodesVersionCompatibility['incompatibleNodes'];
  nodesInfoRequestError?: NodesVersionCompatibility['nodesInfoRequestError'];
}
```
**How to test this:**

The easiest way to test that we're getting information about a request error is by interrupting elasticsearch while Kibana is still running:
1. start es: `yarn es snapshot --license trial`
2. configure logging for the `status` logger:
```
// kibana.yml
logging:
  appenders:
    console_appender:
      type: console
      layout:
        type: pattern
        highlight: true
        pattern: "[%date][%level][%logger][%meta]---%message"
  loggers:
    - name: status
      level: debug
      appenders: [console_appender]
```
3. start Kibana: `yarn start`
4. Wait for Kibana to start up and take a note of the logs for the `status` logger. e.g:
```
[2021-05-13T09:10:35.359-07:00][DEBUG][status][{"kibana":{"status":{"level":"available","summary":"All services are available"}}}]---Recalculated overall status
```
5. stop es
6. Observe the error message (`connect ECONNREFUSED 127.0.0.1:9200`) and more information we see in the `status` logger logs meta (See below for a "prettified version of the `meta` property that contains the error details). e.g.:
```
2021-05-13T09:11:56.275-07:00][DEBUG][status][{"kibana":{"status":{"level":"critical","summary":"[elasticsearch]: Unable to retrieve version information from Elasticsearch nodes. connect ECONNREFUSED 127.0.0.1:9200","meta":{"affectedServices":{"elasticsearch":{"level":"critical","summary":"Unable to retrieve version information from Elasticsearch nodes. connect ECONNREFUSED 127.0.0.1:9200","meta":{"warningNodes":[],"incompatibleNodes":[],"nodesInfoRequestError":{"name":"ConnectionError","meta":{"body":null,"statusCode":null,"headers":null,"meta":{"context":null,"request":{"params":{"method":"GET","path":"/_nodes","body":null,"querystring":"filter_path=nodes.*.version%2Cnodes.*.http.publish_address%2Cnodes.*.ip","headers":{"user-agent":"elasticsearch-js/8.0.0-canary.4 (darwin 20.3.0-x64; Node.js v14.16.1)","x-elastic-product-origin":"kibana","x-elastic-client-meta":"es=8.0.0p,js=14.16.1,t=8.0.0p,hc=14.16.1"},"timeout":30000},"options":{},"id":250},"name":"elasticsearch-js","connection":{"url":"http://localhost:9200/","id":"http://localhost:9200/","headers":{},"deadCount":4,"resurrectTimeout":1620922795408,"_openRequests":1,"status":"dead","roles":{"master":true,"data":true,"ingest":true,"ml":false}},"attempts":3,"aborted":false}}}}}}},"detail":"See the status page for more information"}}}]---Recalculated overall status
```
7. Restart es
8. Observe that the logs for the `status` logger again report that all services are "available":
```
[2021-05-13T09:14:09.129-07:00][DEBUG][status][{"kibana":{"status":{"level":"available","summary":"All services are available"}}}]---Recalculated overall status
```


<details>
  <summary>Prettified version of the `status` logger `meta` field when es is interrupted:</summary>

```
{
  "kibana": {
    "status": {
      "level": "critical",
      "summary": "[elasticsearch]: Unable to retrieve version information from Elasticsearch nodes. connect ECONNREFUSED 127.0.0.1:9200",
      "meta": {
        "affectedServices": {
          "elasticsearch": {
            "level": "critical",
            "summary": "Unable to retrieve version information from Elasticsearch nodes. connect ECONNREFUSED 127.0.0.1:9200",
            "meta": {
              "warningNodes": [],
              "incompatibleNodes": [],
              "nodesInfoRequestError": {
                "name": "ConnectionError",
                "meta": {
                  "body": null,
                  "statusCode": null,
                  "headers": null,
                  "meta": {
                    "context": null,
                    "request": {
                      "params": {
                        "method": "GET",
                        "path": "/_nodes",
                        "body": null,
                        "querystring": "filter_path=nodes.*.version%2Cnodes.*.http.publish_address%2Cnodes.*.ip",
                        "headers": {
                          "user-agent": "elasticsearch-js/8.0.0-canary.4 (darwin 20.3.0-x64; Node.js v14.16.1)",
                          "x-elastic-product-origin": "kibana",
                          "x-elastic-client-meta": "es=8.0.0p,js=14.16.1,t=8.0.0p,hc=14.16.1"
                        },
                        "timeout": 30000
                      },
                      "options": {},
                      "id": 250
                    },
                    "name": "elasticsearch-js",
                    "connection": {
                      "url": "http://localhost:9200/",
                      "id": "http://localhost:9200/",
                      "headers": {},
                      "deadCount": 4,
                      "resurrectTimeout": 1620922795408,
                      "_openRequests": 1,
                      "status": "dead",
                      "roles": {
                        "master": true,
                        "data": true,
                        "ingest": true,
                        "ml": false
                      }
                    },
                    "attempts": 3,
                    "aborted": false
                  }
                }
              }
            }
          }
        }
      },
      "detail": "See the status page for more information"
    }
  }
}
```

</details>

### Checklist

Delete any items that are not applicable to this PR.


### For maintainers
